Add Support for macOS

dcvz opened the pull request:
This PR adds support for macOS! We aimed to keep Apple specific stuff behind check in order to not break Linux (although that's already currently broken in `zelda64`.

After talking to @Kenix3 -- the idea is to continue our work in the mainline. We've still got a couple of more improvements planned but being in the mainline will make our work easier.

This branch was put together by @crowell (on Discord) and I ✨

crowell committed:
add more makefile changes

KiritoDv committed:
Fixed soh filters

crowell committed:
almost ready

crowell committed:
more updates

crowell committed:
update

crowell committed:
update

dcvz committed:
Update Makefiles to handle both platforms

dcvz committed:
Allow for overriding the CXX and CC executables

dcvz committed:
Restore original structure while supporting custom CXX flags

dcvz committed:
Remove some platform specific libs

dcvz committed:
Dynamic target name

dcvz committed:
Make X11 paths package-agnostic

dcvz committed:
Remove changes to `gfx_opengl.cpp`

crowell committed:
Merge pull request #1 from crowell/chore/update-make

Chore/update make

crowell committed:
Use OpenGL2 on MacOS instead of OpenGL3

crowell committed:
make it actually render something

crowell committed:
render at least the first texture, still need to figure out the second
one

dcvz committed:
Let’s use OpenGL 3 again

crowell committed:
maybe this works to get the right texture? link's eyes still look off a bit

crowell committed:
did this work?

crowell committed:
set the platform to macos

crowell committed:
actual numbers are right, but logic is ugly XXX/TODO, i know

dcvz committed:
A bit of cleanup

dcvz committed:
Revert unneeded changes

dcvz committed:
Remove GL_CHECK

crowell committed:
add zlib to ldflags for ZAPDUtils

dcvz committed:
Merge remote-tracking branch 'upstream/zelda64' into zelda64_mac

dcvz committed:
Fix issues with z64 branch

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

# Conflicts:
#	soh/Makefile

crowell committed:
use an std::map instead of a giant array

dcvz committed:
three point filter fix (#2)

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

# Conflicts:
#	soh/soh/OTRGlobals.cpp

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Fix mac compilation

crowell committed:
fix audio for 64 bit

crowell committed:
revert audio heap size, keep bigger pools

dcvz committed:
Add more Apple specific checks to our modifications

ghost commented:
Thanks for your hard work. When I try to build on M1 Mac I get this.

```
~/github/ShipwrightMac/soh$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../libultraship
c++ -Wall -Wextra -Werror -Wno-unused-variable -Wno-unused-parameter -Wno-unused-function -Wno-parentheses -Wno-narrowing -Wno-missing-field-initializers -Wno-error=multichar -std=c++20 -D_GNU_SOURCE -DENABLE_OPENGL -DSPDLOG_ACTIVE_LEVEL=0 -Wno-deprecated-enum-enum-conversion -g -D_DEBUG -MMD -framework OpenGL -D_THREAD_SAFE -I/opt/homebrew/include/SDL2 -I/opt/homebrew/Cellar/glew/2.2.0_1/include -O0 -I../ZAPDTR/ZAPDUtils -I../StormLib/src -Ilibultraship/Lib/Fast3D/U64 -Ilibultraship/Lib/spdlog -Ilibultraship/Lib/spdlog/include -Ilibultraship/Lib/ImGui -Ilibultraship -c libultraship/Factories/CutsceneFactory.cpp -o build/libultraship/Factories/CutsceneFactory.o
clang: error: -framework OpenGL: 'linker' input unused [-Werror,-Wunused-command-line-argument]
make[1]: *** [build/libultraship/Factories/CutsceneFactory.o] Error 1
make: *** [all] Error 2
```

crowell commented:
@BountyChocolate123456 you have to set CC and CXX to gcc-12 and g++-12

(for example in macports

```sh
export CXX=g++-mp-devel
export CC=gcc-mp-devel
```

dcvz committed:
Add building instructions for macOS

dcvz commented:
@BountyChocolate123456 I have added instructions to `BUILDING.md` on how to build :)

ghost commented:
@dcvz Thanks you very much. It works on arm64!! There are tiny audio issues but I think that is due to the 64 bit.

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,30 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
+cmake --build external/StormLib/build
+cp external/StormLib/build/libstorm.a external
+# Copy glew lib to external (path might be different depending on how you installed it)
+cp /usr/local/lib/libGLEW.a external
```
```suggestion
cp `brew --prefix`/lib/libGLEW.a external
```

On M1 libGLEW is available at /opt/homebrew/lib/libGLEW.a

crowell replied:
this step is likely not needed, it's done via pkg-config to find the right .a file

BountyChocolate123456 replied:
Yes, you are correct. I skipped this step and it still compiled

dcvz replied:
True. Will remove this step!

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,30 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
 
+## macOS
+
+1. Requires `gcc@12, sdl2, libpng, glew, spdlog, pulseaudio, x11` (can be installed via brew, etc)
```
I could not find `x11` via brew so I install `libx11`

dcvz replied:
That's correct but it's still x11, will leave this one as is. Not sure if it's called libx11 on other package managers.

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,30 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
+# Extract the assets/Compile the exporter/Run the exporter
+make setup -j8 DEBUG=0 CC=gcc-12 CXX=g++-12
+# Compile the code
+make -j8 DEBUG=0 CC=gcc-12 CXX=g++-12
```
```suggestion
make -j8 DEBUG=0 CC=gcc-12 CXX=g++-12

# Start soh.
./soh.*
```

On M1 it is "./soh.arm64" so the wildcard should start it on Intel Mac too?

dcvz replied:
It would be soh.x86_64 on an Intel Mac

dcvz committed:
Remove unecessary step from building instructions

ghost commented:
@dcvz @crowell I played up to receiveing Sarias song!! No issues!! Great Work!!
Does Rumble work for you? as it didn't work for me.

BountyChocolate123456 commented on libultraship/libultraship/SohImGuiImpl.cpp:
```diff
@@ -1188,8 +1188,12 @@ namespace SohImGui {
             ImGui::Text("Platform: Windows");
+#else
+#ifdef __APPLE__
+            ImGui::Text("Platform: MacOS");
```
```suggestion
            ImGui::Text("Platform: macOS");
```

BountyChocolate123456 commented on soh/.gitignore:
```diff
@@ -18,6 +18,7 @@ notes/
 baserom/
 docs/doxygen/
 *.elf
+*.arm64
```
```suggestion
*.arm64
*.x86_64
```

Do you need to add *.x86_64 for Intel builds?

BountyChocolate123456 commented on BUILDING.md:
````diff
@@ -46,6 +46,28 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
+1. Requires `gcc@12, sdl2, libpng, glew, spdlog, pulseaudio, x11` (can be installed via brew, etc)
+```bash
+# Clone the repo
+git clone git@github.com:HarbourMasters/ShipWright.git
````
```suggestion
git clone https://github.com/HarbourMasters/Shipwright.git
```
Not everyone has ssh keys setup

qurious-pixel commented on soh/Makefile:
```diff
@@ -53,27 +73,60 @@ INC_DIRS := $(addprefix -I, \
 	z \
 	pthread \
 	atomic \
-	SDL2 \
```
SDL2 and GLEW were removed from LDLIBS from linux. 
Darwin uses `$(shell sdl2-config --libs) $(shell pkg-config --libs glew)`
Linux should have similar or add SDL2/GLEW back
```
ifeq ($(UNAME), Linux) #LINUX
LDLIBS += \
```

dcvz replied:
Ah yes! It should be added below in the Linux only section

dcvz committed:
Add missing SDL2 & GLEW to Linux LDLIBS

dcvz committed:
Update BUILDING.md

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

dcvz committed:
Update soh/.gitignore to include other arch binaries

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

dcvz committed:
Use right platform name for debugging window

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

BountyChocolate123456 commented on ZAPDTR/ZAPD/ZRom.cpp:
```diff
@@ -5,8 +5,10 @@
 #include "yaz0/yaz0.h"
 
 #ifndef _MSC_VER
```
```suggestion
#ifdef __linux__
```

BountyChocolate123456 replied:
Can you replace with `#ifdef __linux__`?

BountyChocolate123456 commented on libultraship/libultraship/Lib/Fast3D/gfx_sdl2.cpp:
```diff
@@ -120,8 +122,10 @@ static void set_fullscreen(bool on, bool call_callback) {
 
 static uint64_t previous_time;
 #ifndef __linux__
```
Can you replace these 2 with `#ifdef _WIN32`?

BountyChocolate123456 commented on libultraship/libultraship/Lib/Fast3D/gfx_sdl2.cpp:
```diff
@@ -135,7 +139,9 @@ static void gfx_sdl_init(const char *game_name, bool start_in_fullscreen) {
     SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
 
 #ifndef __linux
+#ifndef __APPLE__
```
Can you replace these 2 with #ifdef _WIN32?

iChantz commented:
When I try to extract and run the exporter i.e. `make setup -j8 DEBUG=0 CC=gcc-12 CXX=g++-12
`
I get the error `g++-12: error: unrecognized command-line option '-m32'`

Any thoughts? On an M1

crowell commented:
@iChantz you likely have the wrong branch checked out. `-m32` in this branch is ONLY ever called on linux platform. let's please leave this for code review and not for technical support, you can ask in the discord for help.

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

# Conflicts:
#	ZAPDTR/Makefile
#	libultraship/Makefile
#	soh/Makefile
#	soh/src/code/audio_load.c

dcvz committed:
Fix stormlib on macos (arm64)

dcvz commented on StormLib/src/SFileAddFile.cpp:
```diff
@@ -100,7 +100,7 @@ static int FillWritableHandle(
 
     // Initialize the file time, CRC32 and MD5
-    assert(sizeof(hf->hctx) >= sizeof(hash_state));
+    //assert(sizeof(hf->hctx) >= sizeof(hash_state));
```
I don't fully understand the changes done to StormLib - but now `[HASH_STATE_SIZE]` is hardcoded, it doesn't seem to be correct for macos(arm64)? -- commenting this out doesn't seem to break anything for either platform.

dcvz replied:
@NEstelami perhaps you have some insights here? I believe you were the one that made the changes to StormLib, right?

NEstelami replied:
For some reason I didn't get a notification for this. There were only two real changes to StormLib - an optimization to make generating the OTRs several seconds faster, and additional checks when files are looked up in order to make `\` and `/` interchangeable. I don't think I ever made any changes to the hash state size or anything like that.

dcvz committed:
Simplify some of the ifdef checks

dcvz committed:
Revert an older no longer necessary fix

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,28 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
+cd ShipWright
+# Copy the baserom to the OTRExporter folder
+cp <path to your ROM> OTRExporter
+# Clone and build StormLib
```
No longer required

dcvz committed:
Remove remaining unecessary deviations

dcvz committed:
Update building instructions after StormLib changes

ghost commented:
This might be a good future enhancement https://github.com/openglonmetal/MGL as my M1 Pro struggles to hit 120fps sometimes. I don't know if the lib is mature enough though.

dcvz commented:
@BountyChocolate123456 our first goal was to get something playable. Our main objective is to create a metal renderer — this is something we’re already working on :)

Edit: thanks for the link though! Might prove helpful ✨

ghost commented:
@dcvz @crowell awesome! Thanks for your hard work on this!!! I tried to port it myself to macOS but couldn't so I'm definetely not complaigning :)

dcvz committed:
Feature: Use OpenGL 4.1 (#1)

dcvz committed:
Further tweak the BUILDING

dcvz committed:
Tidy up

dcvz committed:
reword -j message

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

# Conflicts:
#	soh/src/code/audio_load.c

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Add Jenkins CI Support (#2)

dcvz commented on Jenkinsfile:
```diff
@@ -113,5 +113,33 @@ pipeline {
                 }
             }
         }
+        stage ('Build macOS') {
```
For merging purposes this just verifies that building works and does not do any distribution.
This will come in a later PR once this is merged into mainline

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,24 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
+cd soh
+# Extract the assets/Compile the exporter/Run the exporter
+# -jX defines number of cores to use for compilation - lower or remove entirely if having issues
+make setup -j8 DEBUG=0 CC=gcc-12 CXX=g++-12
```
-jx no longer works for me for make setup. Does it work for you?

dcvz replied:
It does! But no code that we change would affect that, that is a parameter to the `make` command

BountyChocolate123456 replied:
Thanks, it just worked now :-S

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Fix type issues

crowell committed:
add target <appbundle> and <filledappbundle>

add makefile targets to create an .app
`filledappbundle` creates the target with the .otr included

this should perhaps be moved to Application Support though

crowell committed:
pull gcc's rpath from otool output

crowell committed:
move make target to the end so it's not default

dcvz committed:
Add Jenkins and make exe in par with other platforms

dcvz committed:
Actually save build artefacts

dcvz committed:
Fix artefact path

BountyChocolate123456 commented on BUILDING.md:
```diff
@@ -46,6 +46,24 @@ make setup -j$(nproc) OPTFLAGS=-O2 DEBUG=0
 
+## macOS
+
+1. Requires `gcc@12, sdl2, libpng, glew, x11` (can be installed via brew, etc)
```
```suggestion
1. Requires `gcc@12, sdl2, libpng, glew, x11 dylibbundler` (can be installed via brew, etc)
```

dcvz replied:
Updated the building instructions!

dcvz committed:
Remove x11 mentions and linking (not used)

dcvz committed:
Update building instructions for generating app

Kenix3 commented on libultraship/libultraship/Audio.cpp:
```diff
@@ -15,7 +15,7 @@ namespace Ship
 		int dataSize = reader->ReadInt32();
 
-		for (uint32_t i = 0; i < dataSize; i++)
+		for (int i = 0; i < dataSize; i++)
```
```suggestion
		for (size_t i = 0; i < dataSize; i++)
```

Kenix3 commented on libultraship/libultraship/Audio.cpp:
```diff
@@ -24,15 +24,15 @@ namespace Ship
 		int loopStateCnt = reader->ReadUInt32();
 
-		for (uint32_t i = 0; i < loopStateCnt; i++)
+		for (int i = 0; i < loopStateCnt; i++)
```
```suggestion
		for (size_t i = 0; i < loopStateCnt; i++)
```

Kenix3 commented on libultraship/libultraship/Audio.cpp:
```diff
@@ -24,15 +24,15 @@ namespace Ship
 		int bookSize = reader->ReadInt32();
 
-		for (uint32_t i = 0; i < bookSize; i++)
+		for (int i = 0; i < bookSize; i++)
```
```suggestion
		for (size_t i = 0; i < bookSize; i++)
```

dcvz replied:
so book size is coming from `int bookSize = reader->ReadInt32()` which is already saving it to an `int`. Comparing it here as `uint32_t` makes the type checker think you're comparing two wrong types. I'll update the places where we do `ReadInt32();` to actually store as `uint32_t` since that's what it returns.

Kenix3 replied:
So size_t isn't just a size. It's not just an unsigned integer either. It's actually just meant as a _size in bytes._ The loop counters here appear to be reading in bytes based on the value, thus size_t fits.

It may be this is inconsistent and more places need to become size_t as well.

Kenix3 reviewed (changes requested):
Thanks for your hard work on this! I only noticed one issue.

crowell committed:
use appsupport directory

dcvz committed:
Add new app icon

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Update target to match macOS types

dcvz committed:
Update more audio types

dcvz commented:
@Kenix3 I've sorted out the audio types now

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

crowell committed:
fix null deref in Audio_PlayFanfare

dcvz committed:
Merge branch 'zelda64' into zelda64_mac

dcvz committed:
Remove old import from z64

louist103 commented on libultraship/libultraship/SohImGuiImpl.cpp:
```diff
@@ -1323,8 +1327,12 @@ namespace SohImGui {
 
 #ifdef _WIN32
             ImGui::Text("Platform: Windows");
+#else
+#ifdef __APPLE__
+            ImGui::Text("Platform: macOS");
 #else
             ImGui::Text("Platform: Linux");
+#endif
```
`elif`?

louist103 reviewed (changes requested):
Literally one thing then good

crowell committed:
address final nit with apple ifdefs

dcvz commented:
@Kenix3 this is ready to be merged!
